libgcc/config/arc/linux-unwind.h

```diff
@@ -120,10 +120,10 @@ arc_fallback_frame_state (struct _Unwind_Context *context,
 	= ((_Unwind_Ptr) &(regs[i])) - new_cfa;
     }
 
-  fs->regs.reg[31].how = REG_SAVED_VAL_OFFSET;
-  fs->regs.reg[31].loc.offset = ((_Unwind_Ptr) (regs[ret])) - new_cfa;
+  fs->regs.reg[29].how = REG_SAVED_VAL_OFFSET;
+  fs->regs.reg[29].loc.offset = ((_Unwind_Ptr) (regs[ret])) - new_cfa;
 
-  fs->retaddr_column = 31;
+  fs->retaddr_column = 29;
```

Comment:
In ARC700, ilink is a privilege mode register, you can't write to it.
In ARCHS it is also not available as a GPR - how have you tested this patch ?

Reply:
So this fix might be needed for some cases.
The routine above is called nanosleep_nocancel - does this mean it is supposed to NOT ne cancelled.
At any rate see if you need to add cfi pseudo ops to this routine, that should be the first fix

Reply:
Also this is all good analysis and such, but what exactly is the motivation of this change, tracking uClibc failures ? That was something closed a long time ago, whats the sudden interest.

Reply:
In uClibc nanosleep call is nanosleep_nocancel wrapped by LIBC_CANCEL_ASYNC/LIBC_CANCEL_RESET pair. That is why you see nanosleep_nocancel in disassembler.

Reply:
IIUC adding `.cfi` might be necessary: I don't understand why they are not added for `st.aw` and `ld.ab` calls. 
However adding `.cfi` is sill not enough for proper unwind: array contents at blink position should not be rewritten.

Reply:
> Also this is all good analysis and such, but what exactly is the motivation of this change, tracking uClibc failures ? That was something closed a long time ago, whats the sudden interest.

The reason to investigate that was customer issue on arc700 and pthread cancellation. Finding that archs is also impacted was a byproduct of that investigation.